stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/EntityBodyHandler.java

```diff
@@ -186,11 +188,22 @@ public static BlobDataSource constructBlobDataSource(BStruct entityStruct) throw
      */
     public static BJSON constructJsonDataSource(BStruct entityStruct) {
         try {
+            BJSON jsonData;
             Channel byteChannel = getByteChannel(entityStruct);
             if (byteChannel == null) {
                 return null;
             }
-            BJSON jsonData = new BJSON(byteChannel.getInputStream());
+            String contentTypeValue = HeaderUtil.getHeaderValue(entityStruct, HttpHeaderNames.CONTENT_TYPE.toString());
+            if (contentTypeValue != null && !contentTypeValue.isEmpty()) {
+                String charsetValue = MimeUtil.getContentTypeParamValue(contentTypeValue, CHARSET);
+                if (charsetValue != null && !charsetValue.isEmpty()) {
+                    jsonData = new BJSON(byteChannel.getInputStream(), charsetValue);
```

Comment:
As per the constructor `BJSON(InputStream in, String schema, String charset)`, charset should be the third argument.